oak/server/rust/oak_runtime/src/channel.rs

```diff
@@ -0,0 +1,255 @@
+/// when adding `ChannelRef`s into `Message`s.
+#[derive(Clone)]
+pub enum ChannelEither {
+    Writer(ChannelWriter),
+    Reader(ChannelReader),
+}
+
+impl Channel {
+    /// Thread safe method that returns true when there is no longer at least one reader and one
+    /// writer.
+    fn is_orphan(&self) -> bool {
+        self.readers.load(SeqCst) == 0 || self.writers.load(SeqCst) == 0
+    }
+}
+
+/// Creates a new `ChannelWriter` and `ChannelReader` that reference the same underlying `Channel`.
+pub fn new() -> (ChannelWriter, ChannelReader) {
+    let chan = Arc::new(Channel {
+        readers: AtomicUsize::new(1),
+        writers: AtomicUsize::new(1),
+        messages: RwLock::new(VecDeque::new()),
+    });
+    let writer = ChannelWriter(ChannelRef::from_arc(chan.clone()));
+    let reader = ChannelReader(ChannelRef::from_arc(chan.clone()));
+    (writer, reader)
+}
+
+impl ChannelRef {
+    /// Internal method to take a thread-safe reference-counting pointer to a `Channel`.
+    fn from_arc(arc: Arc<Channel>) -> ChannelRef {
+        ChannelRef(arc)
+    }
+
+    /// Attempt to apply a function that takes a `&mut VecDeque<Message>` to the underlying message
+    /// store. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if the underlying channel has been
+    /// orphaned.
+    fn try_write<F, T>(&self, f: F) -> Result<T, OakStatus>
+    where
+        F: FnOnce(&mut Messages) -> T,
+    {
+        if self.0.is_orphan() {
+            return Err(OakStatus::ERR_CHANNEL_CLOSED);
+        }
+
+        let mut messages = self.0.messages.write().unwrap();
+        Ok(f(&mut *messages))
+    }
+
+    /// Apply a function that takes a `&mut VecDeque<Message>` to the underlying message
+    /// store.
+    fn write<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&mut Messages) -> T,
+    {
+        let mut messages = self.0.messages.write().unwrap();
+        f(&mut *messages)
+    }
+
+    /// Apply a function that takes a `&VecDeque<Message>` to the underlying message
+    /// store.
+    fn read<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&Messages) -> T,
+    {
+        let messages = self.0.messages.read().unwrap();
+        f(&*messages)
+    }
+
+    /// Decrement the `Channel` writer counter
+    fn remove_writer(&self) {
+        self.0.writers.fetch_sub(1, SeqCst);
+    }
+
+    /// Decrement the `Channel` reader counter
+    fn remove_reader(&self) {
+        self.0.readers.fetch_sub(1, SeqCst);
+    }
+
+    /// Increment the `Channel` writer counter
+    fn add_writer(&self) -> usize {
+        self.0.writers.fetch_add(1, SeqCst)
+    }
+
+    /// Increment the `Channel` reader counter
+    fn add_reader(&self) -> usize {
+        self.0.readers.fetch_add(1, SeqCst)
+    }
+
+    /// Returns true if there are no messages in the channel _and_ if there is not at least one
+    /// writer and one reader.
+    fn is_empty_orphan(&self) -> bool {
```

Comment:
Why is this more useful than just `self.0.is_orphan()`? Is the idea that if the channel is orphan but there is still some data enqueued, we keep enqueuing more data anyways? Can any reader ever get reattached to the channel in the future?

Reply:
It's only used by readers to allow reading from an orphaned channel that still contains data. Maybe I should move this to ChannelReader?